### PR TITLE
Remove construction of JVP/VJP for curry thunks.

### PR DESCRIPTION
**Pull request by pschuh**
In the Differentiation pass, clone these thunks and promote the signature from (A) -> (Float) -> Float to (A) -> (Float) ~> Float.

### CODE REVIEW

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 1]
```diff
@@ -1435,8 +1455,9 @@ void DifferentiableActivityInfo::analyze(DominanceInfo *di,
 
 bool DifferentiableActivityInfo::setVariedIfDifferentiable(
     SILValue value, unsigned independentVariableIndex) {
-  if (!value->getType().isDifferentiable(function.getModule()))
+  if (!value->getType().isDifferentiable(function.getModule())) {
```
Comment (dan-zheng) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: undo addition of braces here?

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 2]
```diff
@@ -5439,48 +5460,90 @@ class Differentiation : public SILModuleTransform {
...
-      autodiff::getNumAutoDiffAssociatedFunctions(
-          adfi->getDifferentiationOrder()))
-    return false;
-  assert(adfi->getNumAssociatedFunctions() == 0 &&
-         "some functions are already filled in but not all of them");
+SILValue ADContext::promoteToAutodiffFunction(
+    SILBuilder &builder, SILLocation loc, SILValue origFnOperand,
+    const llvm::SmallBitVector &parameterIndices, unsigned differentiationOrder,
+    DifferentiationInvoker invoker) {
+  if (auto *ai = dyn_cast<ApplyInst>(origFnOperand)) {
```
[the remark this answers is not in the document]
Comment (pschuh) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Not sure I understand. Could you clarify what you want?

[the remark this answers is not in the document]
Comment (rxwei) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Nvm, I thought the `if` body covered the entire function.

**Review by rxwei — APPROVED** [review]

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 3]
```diff
@@ -1340,6 +1351,13 @@ void DifferentiableActivityInfo::analyze(DominanceInfo *di,
                 setVariedIfDifferentiable(dirRes, i);
             }
           }
+          auto arg = ai->getCallee();
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: It'd be nice not to repeat this logic. Can you merge this with the loop on line 1346?

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 4]
```diff
@@ -1393,9 +1411,11 @@ void DifferentiableActivityInfo::analyze(DominanceInfo *di,
         // Handle indirect results in `apply`.
         if (auto *ai = dyn_cast<ApplyInst>(&inst)) {
           auto checkAndSetUseful = [&](SILValue res) {
-            if (isUseful(res, i))
+            if (isUseful(res, i)) {
               for (auto arg : ai->getArgumentsWithoutIndirectResults())
                 setUsefulIfDifferentiable(arg, i);
+              setUsefulIfDifferentiable(ai->getCallee(), i);
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: I am not sure if this is necessary for this patch. Looks like it's only needed for full currying.

**Review by rxwei — APPROVED** [review]

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by pschuh** [comment]
@swift-ci please test tensorflow macos

**Comment by pschuh** [comment]
@swift-ci please test tensorflow macos

**Pull request merged**